core/rawdb/freezer.go

```diff
@@ -546,3 +547,82 @@ func (f *freezer) freezeRange(nfdb *nofreezedb, number, limit uint64) (hashes []
 
 	return hashes, err
 }
+
+// isLegacyFn takes a freezer entry and determines if it is in an
+// old format and hence needs to be converted.
+type isLegacyFn = func([]byte) (bool, error)
+
+// convertLegacyFn takes a raw freezer entry in an older format and
+// returns it in the new format.
+type convertLegacyFn = func([]byte) ([]byte, error)
+
+// MigrateTable processes the entries in a given table in sequence
+// converting them to a new format if they're of an old format.
+func (f *freezer) MigrateTable(kind string, isLegacy isLegacyFn, convert convertLegacyFn) error {
+	if f.readonly {
+		return errReadOnly
+	}
+	f.writeLock.Lock()
+	defer f.writeLock.Unlock()
+
+	table, ok := f.tables[kind]
+	if !ok {
+		return errUnknownTable
+	}
+	ancientsPath := filepath.Dir(table.index.Name())
+	// Set up new dir for the migrated table, the content of which
+	// we'll at the end move over to the ancients dir.
+	migrationPath := filepath.Join(ancientsPath, "migration")
+	newTable, err := NewFreezerTable(migrationPath, kind, FreezerNoSnappy[kind])
+	if err != nil {
+		return err
+	}
+	batch := newTable.newBatch()
+	var out []byte
+	// Iterate through entries and transform them
+	table.forEach(func(i uint64, blob []byte) error {
+		if i%500000 == 0 {
+			log.Info("Processing legacy elements", "number", i)
+		}
+		legacy, err := isLegacy(blob)
+		if err != nil {
+			return err
+		}
+		out = blob
+		if legacy {
+			out, err = convert(blob)
+			if err != nil {
+				return err
+			}
+		}
+		if err := batch.AppendRaw(i, out); err != nil {
+			return err
+		}
+		return nil
+	})
+	if err := batch.commit(); err != nil {
+		return err
+	}
+	// Release and delete old table files
+	table.releaseFilesAfter(0, true)
+	// Move migrated files to ancients dir.
+	if err := newTable.Close(); err != nil {
+		return err
+	}
+	files, err := ioutil.ReadDir(migrationPath)
+	if err != nil {
+		return err
+	}
+	for _, f := range files {
+		// This will replace the index + table files up to and including the switchover file.
+		if err := os.Rename(filepath.Join(migrationPath, f.Name()), filepath.Join(ancientsPath, f.Name())); err != nil {
```

Comment:
One thing to be aware of, which I find pretty counter-intuitive, is that `f.Name()` doesn't return the canonical filename, but "the name of the file as presented to Open". 
```
	f, _ := os.Open("/tmp/foo")
	fmt.Printf("name: %v", f.Name())
```
prints out 
```
name: /tmp/foo
```
So it looks to me like this shouldn't work, but I haven't tested it, maybe it does?

Reply:
That is counter-intuitive...however in this case it works because `ioutil.ReadDir` returns a list of `fs.FileInfo` which is a bit different than `os.File` struct.

Reply:
Yes, maybe add some comments about the file name? So that next time we won't be confused.

Reply:
The behavior of `Os.Rename` depends on the operator system. We may need to handle at least windows explicitly.

For example, `go-leveldb` uses this code for windows 
https://github.com/syndtr/goleveldb/blob/master/leveldb/storage/file_storage_windows.go#L66

But I never test it :P